Serde v2 (versioning for serialization/deserialization)

felixguendling opened the pull request:
This introduces a new serde module which provides versioning for aggregate types. The main advantage is a unified serialized header for all types `T` derived from `envelope<T, version<X>, compat_version<Y>>` (`X` being the current version of the data structure `T` and `Y` being the earliest version that's capable of handling serialized data of the data structure at hand). The header is structured as follows: 1 byte version, 1 byte compat version, 4 bytes size. So the "overhead" is 6 bytes. This allows us to distinguish 256 unique versions of the data structure which should be sufficient. Using variable length encoding for the size integer was considered but would either require a second pass over the data structure tree to determine the size of an aggregate type (including all its members) or a `placeholder` in `iobuf` with a variable size (currently it is fixed). The serialization function for each type is customizable. In case the 6 byte header is too inefficient or no versioning is required, it's always possible to implement a handwritten read/write/async_read/async_write function. For aggregate types, this can be done as global function as well as member function.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/vectorizedio/redpanda?pullRequest=1078) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/vectorizedio/redpanda?pullRequest=1078) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/vectorizedio/redpanda?pullRequest=1078) it.</sub>

dotnwat commented:
@felixguendling can you sign the CLA?

felixguendling commented:
> @felixguendling can you sign the CLA?

I thought I already did this?
At least I authorized the CLA app and the bot says "Contributor License Agreement is signed." (CI check).

CLAassistant:
> All committers have signed the CLA.

dotnwat commented on src/v/serde/serde.h:
```diff
@@ -0,0 +1,171 @@
+    using Type = std::decay_t<T>;
+    if constexpr (has_serde_async_read<Type>) {
+        auto const t = std::make_shared<Type>();
+        return t->serde_async_read(in).then([t]() { return *t; });
```
there are a couple ways to handle this, but I don't think we need shared pointer (btw seastar has ss::lw_shared_ptr that should always be used: https://makedist.com/posts/2020/08/02/smart-pointers-in-seastar/).

```
auto t = std::make_unique<Type>();
auto p = t.get();
return p->method().finally([t=std::move(t)]{}); // maybe c++20 does left-to-right guaranteed
```

```
return ss::do_with(Type{}, [](Type& t) {
   return t.method().then([&t] {
      return std::move(t);
   });
});
```

```
Type t;
co_await t.serde_async_read(in);
co_return t;
```

dotnwat commented on src/v/serde/serde.h:
```diff
@@ -0,0 +1,171 @@
+    auto t = Type();
+
+    if constexpr (has_serde_read<Type>) {
+        t.read(in);
```
make normal read have all the versioning enforcement by default

dotnwat commented on src/v/serde/serde.h:
```diff
@@ -0,0 +1,171 @@
+        auto const version = read<version_t>(in);
+        auto const compat_version = read<version_t>(in);
+        auto const size = read<typename Type::envelope_size_t>(in);
+        (void)version; // could be used to drop messages from old versions
```
[[maybe_unused]]

senior7515 commented:
@Lazin might want to take a peek too 

cc: @graphcareful and @BenPope

dotnwat commented:
Please make sure to get your reviews in in the next couple days before Felix makes another pass

dotnwat commented:
@felixguendling hey I re-opened this PR so other could submit reviews.

felixguendling commented:
I cleaned up the commit history, added support for more types, used signed integer types where possible, and added version information as parameter for custom read member functions.

What are the rules regarding copyright? Some files in the code based have the copyright preamble, some (e.g. append_entries_buffer.cc) don't. Is there some rule on when to use it?

senior7515 commented:
> What are the rules regarding copyright? Some files in the code based have the copyright preamble, some (e.g. append_entries_buffer.cc) don't. Is there some rule on when to use it?

all files should have a copyright preamble. it's a bug if not.

senior7515 commented:
I'll try to review this this thursday, but this is looking amazing. I'd be curious to see how this compares with your prev benchs too -  https://cista.rocks/#serialization

dotnwat commented on src/v/serde/serde.h:
```diff
@@ -0,0 +1,260 @@
+            t.serde_write(out);
+        } else {
+            envelope_for_each_field(t, [&out](auto& f) { write(out, f); });
+        }
```
very cool!

dotnwat commented on src/v/serde/serde.h:
```diff
@@ -0,0 +1,260 @@
+          reinterpret_cast<char const*>(&size),
+          sizeof(typename Type::envelope_size_t));
+    } else if constexpr (std::is_enum_v<Type>) {
+        write(out, static_cast<std::underlying_type_t<Type>>(t));
```
is there any way to require that an an enum type has a fixed size. in particular I think both `enum` and `enum class` will match this case and AFAIK both do not have specified fixed size types.

right now we effectively use code review to check that any enum is an `enum class : intXX_t` but I'm hoping we can enforce that some how or use some pattern for making it harder to miss / make a mistake.

looks like maybe is_scoped_enum combined with a checking the underlying type being one of int8_t, int16_t, etc... might work.

dotnwat replied:
maybe remove enum entirely from the set of types we handle automatically

felixguendling replied:
enum support removed because wire size would be platform dependent

dotnwat commented on src/v/serde/serde.h:
```diff
@@ -0,0 +1,260 @@
+        write(out, static_cast<std::underlying_type_t<Type>>(t));
+    } else if constexpr (std::is_scalar_v<Type>) {
+        auto le_t = ss::cpu_to_le(t);
+        out.append(reinterpret_cast<char const*>(&le_t), sizeof(Type));
```
pedantic: should this be sizeof(le_t)?

dotnwat commented on src/v/serde/serde.h:
```diff
@@ -0,0 +1,260 @@
+        auto le_t = ss::cpu_to_le(t);
+        out.append(reinterpret_cast<char const*>(&le_t), sizeof(Type));
+    } else if constexpr (reflection::is_std_vector_v<Type>) {
+        assert(t.size() <= std::numeric_limits<int32_t>::max());
```
you can use our `vassert` variant

felixguendling replied:
used exception instead

dotnwat commented on src/v/serde/serde.h:
```diff
@@ -0,0 +1,260 @@
+        }
+    } else if constexpr (reflection::is_named_type_v<Type>) {
+        return write(iob, static_cast<typename Type::type>(t));
+    } else if constexpr (reflection::is_ss_bool_v<Type>) {
```
if there are only a couple `reflection::*` things we might consider importing them into `serde::*` since that seems like it is probably the future and then we'll have less coupling.

dotnwat replied:
lets' wait

dotnwat commented on src/v/serde/serde.h:
```diff
@@ -0,0 +1,260 @@
+    } else if constexpr (reflection::is_ss_bool_v<Type>) {
+        write(iob, static_cast<int8_t>(bool(t)));
+    } else if constexpr (std::is_same_v<Type, std::chrono::milliseconds>) {
+        write<int64_t>(iob, t.count());
```
what is the difference between `write<int64_t>(iob, t.count());` here and `write(iob, static_cast<int8_t>(bool(t)));` above in terms of specializing / invoking the correct overload?

dotnwat commented on src/v/serde/serde.h:
```diff
@@ -0,0 +1,260 @@
+          sizeof(typename Type::envelope_size_t));
+    } else if constexpr (std::is_enum_v<Type>) {
+        write(out, static_cast<std::underlying_type_t<Type>>(t));
+    } else if constexpr (std::is_scalar_v<Type>) {
```
I think that `is_scalar_v` will also handle `bool`? in that case, it seems like can't necessarily rely on `bool` having a predictable representation? in this case, it seems like it is going to be dependent which particular overload of `auto le_t = ss::cpu_to_le(t);`is chosen?

dotnwat replied:
related to is_enum case above because is_scalar is true for Type is enum

dotnwat replied:
ss::cpu_to_le(t); may not give predictable behavior in futures version of seastar when given a bool

dotnwat commented on src/v/serde/serde.h:
```diff
@@ -0,0 +1,260 @@
+    auto const compat_version = read<version_t>(in);
+    auto const size = read<typename Type::envelope_size_t>(in);
+
+    if (compat_version > Type::__version) {
```
`if (unlikely(...))`?

dotnwat commented on src/v/serde/serde.h:
```diff
@@ -0,0 +1,260 @@
+          type_str<T>(),
+          static_cast<int>(Type::__version));
+        throw std::system_error{
+          make_error_code(error_codes::version_older_than_compat_version)};
```
can you comment on logging the error here vs adding more context to the exception? I do agree though we do need some way to get contextual information out. also, I think we could probably do this trick here https://github.com/vectorizedio/redpanda/blob/dev/src/v/utils/utf8.h#L29. and with that trick it probably isn't so bad having the log statement.

dotnwat replied:
throw serde_exeception(fmt_with_ctx(ssx::format, "asdlkfjskldjf {} {}", params));

dotnwat commented on src/v/serde/serde.h:
```diff
@@ -0,0 +1,260 @@
+}
+
+template<typename T>
+ss::future<std::decay_t<T>> read_async(iobuf_parser& in) {
```
this function is amazing!

dotnwat reviewed (commented):
amazing! that was a lot of fun to review! it's looking really really good. I didn't get around to applying it to types outside of the test cases you have, but that is next, and I don't think that needs to hold up merging.

dotnwat commented on src/v/serde/envelope.h:
```diff
@@ -0,0 +1,68 @@
+#pragma once
```
failing in ci. we moved recently to clang-format 12

dotnwat commented on src/v/serde/envelope.h:
```diff
@@ -0,0 +1,68 @@
+struct envelope {
+    using envelope_size_t = uint32_t;
+    using value_t = T;
+    static constexpr auto __version = Version::v;
```
maybe something obnoxious like redpanda_serde_version?

dotnwat commented on src/v/serde/envelope.h:
```diff
@@ -0,0 +1,68 @@
+    static constexpr auto const value = std::is_base_of_v<
+      envelope<
+        Type,
+        version<Type::__version>,
```
yes

dotnwat commented on src/v/serde/envelope.h:
```diff
@@ -0,0 +1,68 @@
+template<typename T>
+inline constexpr auto const is_envelope_v = std::conjunction_v<
+  detail::has_compat_attribute<T>,
+  detail::inherits_from_envelope<T>>;
```
1. need this for version, in addition to compat version?
2. going to check if this is still needed
3. perhaps we can statically assert that T does not have members that conflict with our version/compat_version name. this is not a clear thought.

if we choose to keep this I think we need to write some comments here about what's going on.

felixguendling replied:
```cpp
std::conjunction_v<
  detail::has_compat_attribute<T>,
  detail::has_version_attribute<T>>;
```

It now just requires `redpanda_serde_compat_version` + `redpanda_serde_compat_version`. So theoretically, you do not need to inherit from `envelope`. Just add those two attributes to your struct, and you're fine. However, using `envelope` could be more future-proof because we can make changes centrally.

dotnwat commented on src/v/serde/envelope_for_each_field.h:
```diff
@@ -0,0 +1,181 @@
+#pragma once
+
+#include "reflection/arity.h"
```
let's not do this yet

dotnwat commented on src/v/serde/serde.h:
```diff
@@ -0,0 +1,260 @@
+HAS_MEMBER_FN(serde_read, iop, version_t{0}, version_t{0}, size_t{0U})
+HAS_MEMBER_FN(serde_write, iob)
+HAS_MEMBER_FN(serde_async_read, iop, version_t{0}, version_t{0}, size_t{0U})
+HAS_MEMBER_FN(serde_async_write, iob)
```
maybe use concepts here.

dotnwat replied:
static means same translation unit so no linking issues but lets clean this up.

dotnwat commented on src/v/serde/serde.h:
```diff
@@ -0,0 +1,260 @@
+  std::is_same_v<T, iobuf> || std::is_same_v<T, ss::sstring> ||
+  reflection::is_std_optional_v<T>;
+
+using header_t = std::tuple<version_t, version_t, size_t>;
```
consider strong type for version_t

dotnwat commented on src/v/serde/serde.h:
```diff
@@ -0,0 +1,260 @@
+        } else {
+            write(out, false);
+        }
+    }
```
this "trick" is cool, but it's a trick. let's be explicit with the list of types.

BenPope reviewed (commented):
Pretty cool

dotnwat commented on src/v/serde/serde.h:
```diff
@@ -0,0 +1,260 @@
+static auto iob = iobuf{};
+static auto iop = iobuf_parser{std::move(iob)};
+
+HAS_MEMBER_FN(serde_read, iop, version_t{0}, version_t{0}, size_t{0U})
```
Hey ben could you clarify what you meant here? BTW, this code is completely rewritten without macros etc...in the new version so maybe a new review is the easiest thing to do.

BenPope replied:
Just had a quick glance.  Essentially the pattern is:
```cpp
T t;
t.serde_read(args...);
```
I thought it might be useful to have a hook for:
```cpp
auto t{serde_construct<T>(args...);
```
Where `T` no longer requires a default constructor.

felixguendling replied:
Just to clarify: this would replace the (user implemented) `serde_read` member method, right?
So you would implement the global `serde_construct<T>` for your type (maybe in the serde namespace or it could work by ADL if you implement it in the namespace of the struct itself). But the write still happens through the member function serde_write()?

How about the asynchronous variants? I think the idea to eliminate the need for a default ctor is good. But I'm not sure on how to do it asynchronously. Maybe we need a `serde_construct_async<T>(args...) -> future<>` to replace `ss::do_with(T{}, [](T& t) { t.serde_read_async(args...).then(...) };`?

Is this a practical issue? Are there many types (that should go on the wire/disk) that don't have a default ctor and where it's hard/impossible to quickly write one?

BenPope replied:
It might be possible to choose construction via constructor, or aggregate init as fallback 

We have clang-14 / full concepts now, if that simplifies the implementation.

dotnwat commented on src/v/serde/serde.h:
```diff
@@ -0,0 +1,314 @@
+          reinterpret_cast<char const*>(&size), sizeof(serde_size_t));
+    } else if constexpr (std::is_same_v<bool, Type>) {
+        write<int8_t>(out, t);
+    } else if constexpr (std::is_scalar_v<Type> && !std::is_enum_v<Type>) {
```
maybe just std::is_arithmetic since scalar handles stuff like pointers that we don't care about

dotnwat replied:
double check float/double wire encoding

felixguendling replied:
Scalars work all the same:
https://github.com/google/flatbuffers/blob/6ed780dbd50f3fd4f2a12b13f5ef709da5372401/include/flatbuffers/base.h#L390-L396

dotnwat commented on src/v/serde/serde.h:
```diff
@@ -0,0 +1,314 @@
+            out.append(reinterpret_cast<char const*>(&le_t), sizeof(le_t));
+        }
+    } else if constexpr (reflection::is_std_vector_v<Type>) {
+        if (unlikely(t.size() > std::numeric_limits<serde_size_t>::max())) {
```
unlikey and/or [[gnu::cold]] ?

BenPope commented:
GitHub won't let me reply to this inline.

> Just to clarify: this would replace the (user implemented) serde_read member method, right?

I thought of it as an optional extra.

> So you would implement the global serde_construct<T> for your type (maybe in the serde namespace or it could work by ADL if you implement it in the namespace of the struct itself). But the write still happens through the member function serde_write()?

I think that makes sense (I was thinking an ADL extension point).

> How about the asynchronous variants? I think the idea to eliminate the need for a default ctor is good. But I'm not sure on how to do it asynchronously. Maybe we need a serde_construct_async<T>(args...) -> future<> to replace ss::do_with(T{}, [](T& t) { t.serde_read_async(args...).then(...) };?

I hadn't considered that construction could be async, but that makes sense (`serde_construct_async<T>(args...) -> future<T>`).

> Is this a practical issue? Are there many types (that should go on the wire/disk) that don't have a default ctor and where it's hard/impossible to quickly write one?

It might be YAGNI. There were a bunch of types in kafka/protocol that had convenience constructors, but I think in a lot of places the constructors were replaced with a combination of default member init and designated init (to ensure ordering), since they are pretty much just used as DTOs (aggregates).

It probably makes sense to defer this until such a usage comes about.

dotnwat commented:
/gcbrun

dotnwat commented:
@felixguendling the following error is showing up in CI. libc++ doesn't have concepts so I don't think that convenience things like std::convertible_to are not available.

```
Step #2 - "run task": [docker:task] [rp:build] ccache /vectorized/llvm/bin/clang++ -DBOOST_TEST_DYN_LINK -DFMT_LOCALE -DSEASTAR_API_LEVEL=6 -DSEASTAR_DEBUG -DSEASTAR_DEBUG_SHARED_PTR -DSEASTAR_DEFAULT_ALLOCATOR -DSEASTAR_SHUFFLE_TASK_QUEUE -DSEASTAR_TYPE_ERASE_MORE -DSERDE_TEST -DXXH_PRIVATE_API -DZSTD_STATIC_LINKING_ONLY -I../../../src/v -Isrc/v -isystem /vectorized/include -isystem /vectorized/include/hdr -fPIC -march=westmere -fuse-ld=lld -mllvm -inline-threshold=2500 -Wno-unused-lambda-capture -Wno-unused-command-line-argument -stdlib=libc++ -fsanitize=undefined -fsanitize=address -O0 -g2 -gsplit-dwarf -fPIE -std=c++20 -U_FORTIFY_SOURCE -DSEASTAR_SSTRING -Werror=unused-result "-Wno-error=#warnings" -fstack-clash-protection -fsanitize=address -fsanitize=undefined -fno-sanitize=vptr -std=c++20 -MD -MT src/v/serde/test/CMakeFiles/serde_rpunit.dir/serde_test.cc.o -MF src/v/serde/test/CMakeFiles/serde_rpunit.dir/serde_test.cc.o.d -o src/v/serde/test/CMakeFiles/serde_rpunit.dir/serde_test.cc.o -c ../../../src/v/serde/test/serde_test.cc
Step #2 - "run task": [docker:task] [rp:build] In file included from ../../../src/v/serde/test/serde_test.cc:12:
Step #2 - "run task": [docker:task] [rp:build] ../../../src/v/serde/serde.h:90:14: error: expected concept name with optional arguments
Step #2 - "run task": [docker:task] [rp:build]         } -> std::convertible_to<ss::future<>>;
Step #2 - "run task": [docker:task] [rp:build]              ^
Step #2 - "run task": [docker:task] [rp:build] ../../../src/v/serde/serde.h:110:10: error: unknown type name 'HasAsyncWrite'
Step #2 - "run task": [docker:task] [rp:build] template<HasAsyncWrite T>
Step #2 - "run task": [docker:task] [rp:build]          ^
Step #2 - "run task": [docker:task] [rp:build] ../../../src/v/serde/serde.h:111:33: error: unknown type name 'T'
Step #2 - "run task": [docker:task] [rp:build] void write_envelope(iobuf& out, T const& t) {
Step #2 - "run task": [docker:task] [rp:build]                                 ^
Step #2 - "run task": [docker:task] [rp:build] 3 errors generated.
```

BenPope commented:
> @felixguendling the following error is showing up in CI. libc++ doesn't have concepts so I don't think that convenience things like std::convertible_to are not available.

I think the helpers have been added to to libc++ 13.

We could add them with a convenience header, or stick to SFINAE with the void_t detection idiom, but concepts are so much simpler.

dotnwat commented:
> > @felixguendling the following error is showing up in CI. libc++ doesn't have concepts so I don't think that convenience things like std::convertible_to are not available.
> 
> I think the helpers have been added to to libc++ 13.
> 
> We could add them with a convenience header, or stick to SFINAE with the void_t detection idiom, but concepts are so much simpler.

Good idea. Perhaps we can include just the ones we need (e.g. https://github.com/llvm/llvm-project/blob/main/libcxx/include/concepts#L165) and then add a macro that will #error on compiling when its available in libc++.

senior7515 commented:
> > > @felixguendling the following error is showing up in CI. libc++ doesn't have concepts so I don't think that convenience things like std::convertible_to are not available.
> > 
> > 
> > I think the helpers have been added to to libc++ 13.
> > We could add them with a convenience header, or stick to SFINAE with the void_t detection idiom, but concepts are so much simpler.
> 
> Good idea. Perhaps we can include just the ones we need (e.g. https://github.com/llvm/llvm-project/blob/main/libcxx/include/concepts#L165) and then add a macro that will #error on compiling when its available in libc++.

don't we have that macro CONCEPTS or smth like that. when avail, they'll work when not, it will be duck typing :D

BenPope commented:
> Good idea. Perhaps we can include just the ones we need (e.g. https://github.com/llvm/llvm-project/blob/main/libcxx/include/concepts#L165) and then add a macro that will #error on compiling when its available in libc++.

Might be able to make it optional with a library feature or version check, which will keep it working for older compilers when we upgrade.

BenPope commented:
> don't we have that macro CONCEPTS or smth like that. when avail, they'll work when not, it will be duck typing :D

But you can't use the concept for removing something from the overload set if it's MACRO'd away. Hmm, that would prevent it working on compilers without concept support.

BenPope commented:
There's always the range-v3 concept emulation layer, but I'm not opposed to requiring concepts support in the compiler, the library bit is lightweight, at least.

dotnwat commented:
> > > > @felixguendling the following error is showing up in CI. libc++ doesn't have concepts so I don't think that convenience things like std::convertible_to are not available.
> > > 
> > > 
> > > I think the helpers have been added to to libc++ 13.
> > > We could add them with a convenience header, or stick to SFINAE with the void_t detection idiom, but concepts are so much simpler.
> > 
> > 
> > Good idea. Perhaps we can include just the ones we need (e.g. https://github.com/llvm/llvm-project/blob/main/libcxx/include/concepts#L165) and then add a macro that will #error on compiling when its available in libc++.
> 
> don't we have that macro CONCEPTS or smth like that. when avail, they'll work when not, it will be duck typing :D

This is true if you only care about the nicer error messages. You can also use concepts to do SFINAE type stuff. I think Felix is doing the later here.

BenPope commented:
It's a separate issue, anyway, as I think all the SFINAE machinery has been implemented, it's just that concepts are a nice little extra.

felixguendling commented:
If have changed it to a `if constexpr` (which is not worse I think) so it should work now.

dotnwat commented:
> If have changed it to a `if constexpr` (which is not worse I think) so it should work now.

did you push an update to the branch?

dotnwat commented:
/gcbrun

unicomp21 commented:
@dotnwat is there a way to map between this and rust serde?  If I'm thinking correctly, the multilanguage support would have all kinds of nice side effects?

https://docs.rs/serde-generate/0.19.3/serde_generate/

dotnwat commented:
> @dotnwat is there a way to map between this and rust serde? If I'm thinking correctly, the multilanguage support would have all kinds of nice side effects?
> 
> https://docs.rs/serde-generate/0.19.3/serde_generate/

hey @unicomp21 the new serde system is intended to only be used internally by c++ for on disk formats and as internal communication between redpanda nodes. the format itself could be compatible with other languages, but we haven't written any code generators other than for c++.
